### PR TITLE
SOLR-15461: Upgrade Apache Calcite to latest release.

### PR DESCRIPTION
**Pull request by markrmiller**
Almost ready, but there is still a test that fails with an off by one result count.

### CODE REVIEW

**Comment by markrmiller** [comment]
Merging up had brought up another 3-4 failing tests that I'm looking at.

**Comment by markrmiller** [comment]
I'm getting around most of the new issues using RexUtil.expandSearch. Looks like the new In and Between stuff is returning a SEARCH SQL kind and this translates back to the expected. The final test in the Between test case still failing though.

**Comment by thelabdude** [comment]
The test failure is because Calcite's expression reducer rules is removing the second part of the AND statement b/c it looks unnecessary to Calcite. From the debug:
```
rel#718:LogicalFilter.NONE.[](input=RelSubset#699,condition==(CAST($0):VARCHAR, 'XXXX'))
                  call#7692 rule [ReduceExpressionsRule(Filter)]
                    rel#700:LogicalFilter.NONE.[](input=RelSubset#699,condition=AND(=(CAST($0):VARCHAR, 'XXXX'), <>(CAST($0):VARCHAR, 'XXXY')))
                      no parent
```
See how the expression gets "reduced" to `(CAST($0):VARCHAR, 'XXXX'))`. This is because Calcite doesn't know about search engine semantics. Personally, I think the issue here is mapping the `=` operator to a term match in an analyzed text field vs. real equality.

These changes fix the breakage:
```
diff --git a/solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java b/solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java
index 60990c07e19..5a1f9c4a1fa 100644
--- a/solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java
+++ b/solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java
@@ -1167,7 +1167,7 @@ public class TestSQLHandler extends SolrCloudTestCase {
 
     sParams = mapParams(CommonParams.QT, "/sql", "aggregationMode", "facet",
         "stmt", "select str_s, count(*), sum(field_i), min(field_i), max(field_i), "
-            + "cast(avg(1.0 * field_i) as float) from collection1 where (text_t='XXXX' AND NOT (text_t='XXXY')) "
+            + "cast(avg(1.0 * field_i) as float) from collection1 where (text_t LIKE 'XXXX' AND NOT (text_t LIKE 'XXXY')) "
             + "group by str_s order by str_s desc");
 
     tuples = getTuples(sParams, baseUrl);
@@ -1203,7 +1203,7 @@ public class TestSQLHandler extends SolrCloudTestCase {
 
     sParams = mapParams(CommonParams.QT, "/sql", "aggregationMode", "facet",
         "stmt", "select str_s as myString, count(*), sum(field_i) as mySum, min(field_i), max(field_i), "
-            + "cast(avg(1.0 * field_i) as float) from collection1 where (text_t='XXXX' AND NOT (text_t='XXXY')) "
+            + "cast(avg(1.0 * field_i) as float) from collection1 where (text_t LIKE 'XXXX' AND NOT (text_t LIKE 'XXXY')) "
             + "group by str_s order by myString desc");
 
     tuples = getTuples(sParams, baseUrl);
```

**Review by thelabdude — APPROVED** [review]
Looks good!

**Comment by markrmiller** [comment]
I took a look at a flag to pull the change regarding the filter, expression, reduction rule that I removed, but initially I ended up leaving it for a couple of reasons:

* Taking it out actually is what maintains backcompat, at least with regards to what we test on.
* The result of not pulling it currently seems to me to produce incorrect results for the query(s) that no longer pass after update and either keeping the rule or removing it doesn't seem to break or affect any other queries we test.

So rather than add this cruft prematurely, I've gone without, but will add it if it is or becomes apparent to someone that we need it.

File: solr/CHANGES.txt [hunk 1]
```diff
@@ -335,6 +335,8 @@ Other Changes
   collections/collectionName into its state.json.  For many-collection clusters, this is an
   optimization when the cluster status is fetched.  (Nazerke Seidan, David Smiley)
 
+* SOLR-15461: SOLR-15461: Upgrade Apache Calcite to 1.27.0. (Mark Miller, Timothy Potter)
```
Comment (madrob) [inline solr/CHANGES.txt:338]: Do we need to document anything about backwards compatibility? Or call out new features in the ref guide

**Pull request merged**